server/tasks/process_file.py

```diff
@@ -30,32 +34,82 @@ def process_file(serialized_agent: bytes):
     asyncio.run(agent.environment.finalize())
 
 
-@app.task(name="process_file_streaming", track_started=True, bind=True)
-def process_file_streaming(self, serialized_agent: bytes):
-    # Load the agent
-    agent = pickle.loads(serialized_agent)
+@app.task(
+    name="streaming_parent_task", track_started=True, bind=True, serializer="pickle"
+)
+def streaming_parent_task(
+    self, agent: Agent, result_handler: ResultHandler, batch_size: int = 2
+):
+    # Parent job ID is used for input/output topic names
+    parent_job_id = self.request.id
+
+    task = process_file_streaming
+    logger.info(f"Submitting task {task.name} with agent {agent}")
+    input_result = task.delay(agent=agent, parent_job_id=parent_job_id)
+    input_job_id = input_result.id
+    logger.info(f"Task {task.name} submitted with job_id {input_job_id}")
+
+    task = process_streaming_output
+    logger.info(f"Submitting task {task.name}")
+    output_result = task.delay(
+        input_job_id=input_job_id,
+        parent_job_id=parent_job_id,
+        result_handler=result_handler,
+        batch_size=batch_size,
+    )
+    output_job_id = output_result.id
+    logger.info(f"Task {task.name} submitted with job_id {output_job_id}")
+
+    # Store input and output job IDs in parent task metadata
+    # Need to pass state as well otherwise its overwritten to None
+    self.update_state(
+        state=states.STARTED,
+        meta={"input_job_id": input_job_id, "output_job_id": output_job_id},
+    )
+
+    input_job = process_file_streaming.AsyncResult(input_job_id)
+    output_job = process_streaming_output.AsyncResult(output_job_id)
+
+    terminal_statuses = ["SUCCESS", "FAILURE", "REVOKED"]
+
+    while (
+        input_job.status not in terminal_statuses
+        or output_job.status not in terminal_statuses
+    ):
+        time.sleep(1)
```

Comment:
I'm pretty sure this is going to be blocking; I can't remember quite how this works in python, but is there a simple way to release the CPU while we're waiting here?

Reply:
No time.sleep does release the GIL - not blocking. This is the simple way